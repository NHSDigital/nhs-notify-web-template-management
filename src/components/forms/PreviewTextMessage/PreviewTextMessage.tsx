'use client';

import { Radios } from 'nhsuk-react-components';
import { Preview } from '@/molecules/Preview';
import { PreviewMessage } from '@/templates/PreviewMessage';
import { PreviewTextMessageProps } from './PreviewTextMessage.types';
import MarkdownIt from 'markdown-it';

const SMS_MD_OPTS: string[] = [];

export function PreviewTextMessage({
  templateName,
  message,
}: PreviewTextMessageProps) {
  const md = new MarkdownIt('zero').enable(SMS_MD_OPTS);

  const html = md.render(message);

  return (
    <div className='nhsuk-grid-row'>
      <div className='nhsuk-grid-column-two-thirds'>
        <PreviewMessage
          type='Text message'
          templateName={templateName}
          details={{
            heading: 'Who your text message will be sent from',
            text: [
              'Set your text message sender name during onboarding.',
              'If you need to set up a different text message sender name for other messages, contact our onboarding team.',
            ],
          }}
          PreviewComponent={
            <Preview
              preview={[
                {
                  heading: 'Message',
                  value: message,
                },
              ]}
            />
          }
          FormOptionsComponent={
            <Radios id='what-would-you-like-to-do-next' name='choice'>
              <Radios.Radio value='edit'>Edit</Radios.Radio>
              <Radios.Radio value='send'>Send a text message</Radios.Radio>
              <Radios.Radio value='submit'>Submit</Radios.Radio>
            </Radios>
          }
        />
      </div>
    </div>
  );
}
