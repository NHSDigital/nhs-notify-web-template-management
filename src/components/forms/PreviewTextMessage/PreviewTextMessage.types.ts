export type PreviewTextMessageProps = {
  templateName: string;
  message: string;
};
