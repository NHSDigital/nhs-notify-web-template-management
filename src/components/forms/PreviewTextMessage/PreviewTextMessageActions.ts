import { MarkdownItWrapper } from '@utils/markdownit';
import {
  Dependencies,
  IPreviewTextMessageActions,
} from './PreviewTextMessage.types';

export class PreviewTextMessageActions implements IPreviewTextMessageActions {
  private readonly _markdownClient: MarkdownItWrapper;

  constructor(
    _container: Dependencies = { markdownClient: new MarkdownItWrapper() }
  ) {
    this._markdownClient = _container.markdownClient;
  }

  renderMarkdown(value: string) {
    'use server';
    return this._markdownClient.render(value);
  }
}
