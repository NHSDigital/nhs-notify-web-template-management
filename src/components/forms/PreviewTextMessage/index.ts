export * from './PreviewTextMessage';
export * from './PreviewTextMessage.types';
export * from './server-actions';
