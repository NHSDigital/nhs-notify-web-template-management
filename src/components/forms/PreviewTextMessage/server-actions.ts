import { FormState } from '@utils/types';
import { MarkdownItWrapper } from '@utils/markdownit';

export function renderMarkdown(
  value: string,
  markdown = new MarkdownItWrapper()
) {
  return markdown.render(value);
}

export function handleForm(state: FormState, _data: FormData) {
  return { ...state };
}
