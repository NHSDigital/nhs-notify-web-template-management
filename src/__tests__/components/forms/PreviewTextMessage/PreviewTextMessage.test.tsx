import { render, screen } from '@testing-library/react';
import { PreviewTextMessage, renderMarkdown } from '@forms/PreviewTextMessage';

jest.mock('@forms/PreviewTextMessage/server-actions');

describe('Preview sms form renders', () => {
  it('matches snapshot', () => {
    const container = render(
      <PreviewTextMessage templateName='test-template-sms' message='message' />
    );

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('renders component correctly', () => {
    render(
      <PreviewTextMessage
        templateName='test-template-sms'
        message='sms message body'
      />
    );

    expect(screen.getByTestId('preview-sms-form__radios-edit')).toHaveAttribute(
      'value',
      'edit'
    );

    expect(screen.getByTestId('preview-sms-form__radios-send')).toHaveAttribute(
      'value',
      'send'
    );

    expect(
      screen.getByTestId('preview-sms-form__radios-submit')
    ).toHaveAttribute('value', 'submit');
  });

  it('should should render message with markdown', () => {
    const renderMock = jest.mocked(renderMarkdown);

    renderMock.mockReturnValue('Rendered via MD');

    const message = 'sms message body';
    render(
      <PreviewTextMessage templateName='test-template-sms' message={message} />
    );

    expect(renderMock).toHaveBeenCalledWith(message);

    expect(screen.getByTestId('preview__content-0')).toHaveTextContent(
      'Rendered via MD'
    );
  });
});
