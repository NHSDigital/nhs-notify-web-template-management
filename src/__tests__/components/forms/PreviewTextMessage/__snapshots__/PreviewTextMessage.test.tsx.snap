// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview sms form renders matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <h1
        data-testid="preview-message__heading"
      >
        <span
          class="nhsuk-caption-l"
          data-testid="preview-message__heading-caption"
        >
          Text message template
        </span>
        test-template-sms
      </h1>
      <details
        class="nhsuk-details"
      >
        <summary
          class="nhsuk-details__summary"
          data-testid="preview-message-details__heading"
        >
          <span
            class="nhsuk-details__summary-text"
          >
            Who your text message will be sent from
          </span>
        </summary>
        <div
          class="nhsuk-details__text"
          data-testid="preview-message-details__text"
        >
          <p>
            Set your text message sender name during onboarding.
          </p>
          <p>
            If you need to set up a different text message sender name for other messages, contact our onboarding team.
          </p>
        </div>
      </details>
      <table
        class="nhsuk-table preview nhsuk-u-margin-bottom-4"
      >
        <tbody
          class="nhsuk-table__body"
        >
          <tr
            class="nhsuk-table__row"
            role="row"
          >
            <td
              class="nhsuk-table__cell"
              role="cell"
            >
              <div
                class="preview__heading"
                data-testid="preview__heading-0"
              >
                Message
              </div>
            </td>
            <td
              class="nhsuk-table__cell"
              role="cell"
            >
              <div
                class="preview__content"
                data-testid="preview__content-0"
              />
            </td>
          </tr>
        </tbody>
      </table>
      <form>
        <div
          class="nhsuk-form-group"
        >
          <fieldset
            class="nhsuk-fieldset"
          >
            <legend
              class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m"
              data-testid="preview-message-form__legend"
            >
              What would you like to do next with this template?
            </legend>
            <div
              class="nhsuk-form-group"
            >
              <div
                class="nhsuk-radios"
                id="what-would-you-like-to-do-next"
              >
                <div
                  class="nhsuk-radios__item"
                >
                  <input
                    aria-labelledby="what-would-you-like-to-do-next-1--label"
                    class="nhsuk-radios__input"
                    data-testid="preview-sms-form__radios-edit"
                    id="what-would-you-like-to-do-next-1"
                    name="choice"
                    type="radio"
                    value="edit"
                  />
                  <label
                    class="nhsuk-label nhsuk-radios__label"
                    for="what-would-you-like-to-do-next-1"
                    id="what-would-you-like-to-do-next-1--label"
                  >
                    Edit
                  </label>
                </div>
                <div
                  class="nhsuk-radios__item"
                >
                  <input
                    aria-labelledby="what-would-you-like-to-do-next-2--label"
                    class="nhsuk-radios__input"
                    data-testid="preview-sms-form__radios-send"
                    id="what-would-you-like-to-do-next-2"
                    name="choice"
                    type="radio"
                    value="send"
                  />
                  <label
                    class="nhsuk-label nhsuk-radios__label"
                    for="what-would-you-like-to-do-next-2"
                    id="what-would-you-like-to-do-next-2--label"
                  >
                    Send a text message
                  </label>
                </div>
                <div
                  class="nhsuk-radios__item"
                >
                  <input
                    aria-labelledby="what-would-you-like-to-do-next-3--label"
                    class="nhsuk-radios__input"
                    data-testid="preview-sms-form__radios-submit"
                    id="what-would-you-like-to-do-next-3"
                    name="choice"
                    type="radio"
                    value="submit"
                  />
                  <label
                    class="nhsuk-label nhsuk-radios__label"
                    for="what-would-you-like-to-do-next-3"
                    id="what-would-you-like-to-do-next-3--label"
                  >
                    Submit
                  </label>
                </div>
              </div>
            </div>
          </fieldset>
        </div>
        <button
          aria-disabled="false"
          class="nhsuk-button"
          data-testid="preview-message-form__button"
          type="submit"
        >
          Continue
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
